src/functional/iterator/transforms/merge_let.py

```diff
@@ -0,0 +1,54 @@
+import eve
+from functional.iterator import ir as itir
+from functional.iterator.transforms.count_symbol_refs import CountSymbolRefs
+
+
+class MergeLet(eve.NodeTranslator):
+    """
+    Merge let-like statements.
+
+    For example transforms::
+
+        (λ(a) → (λ(b) → a+b)(arg1))(arg2)
+
+    into::
+
+        (λ(a, b) → a+b)(arg1, arg2)
+
+    This can significantly reduce the depth of the tree and its readability.
+    """
+
+    def visit_FunCall(self, node: itir.FunCall):
+        node = self.generic_visit(node)
+        if (
+            isinstance(node.fun, itir.Lambda)
+            and isinstance(node.fun.expr, itir.FunCall)
+            and isinstance(node.fun.expr.fun, itir.Lambda)
+        ):
+            outer_lambda = node.fun
+            outer_lambda_args = node.args
+            inner_lambda = node.fun.expr.fun
+            inner_lambda_args = node.fun.expr.args
+            # skip if we have a collision
+            if set(outer_lambda.params) & set(inner_lambda.params):
+                return node
+            # check if the arguments to the outer lambda call use a symbol of the inner lambda
+            # e.g. (λ(a) → (λ(b) → b)(a))(b)
+            ref_counts_outer = CountSymbolRefs.apply(
+                outer_lambda_args, [param.id for param in inner_lambda.params]
+            )
+            if any(ref_count != 0 for ref_count in ref_counts_outer.values()):
```

Comment:
Same, but shorter:
```suggestion
            if any(ref_counts_outer.values()):
```

Reply:
But actually, this should be faster I guess (early exit):
```python
if any(trees.walk_values(outer_lambda_args).if_isinstance(ir.SymbolRef).getattr("id").if_in({p.id for p in inner_lambda.params})):
```

Reply:
I kept the more explicit version, which I find easier to digest. The `walk_values` approach is not able to handle symbol collisions if there are new symbol declarations in `outer_lambda_args`. We might want to consider using unique symbols at some point to use this faster approach.